### PR TITLE
Update 'Populate switch' to support switch-expressions.

### LINKED ISSUE #37639: 'Add missing cases' does not show up for switch expressions
VS 16.3.0 Preview 1.0

```cs
swi[||]tch (((EntityHandle)h).Kind)
{
}
```

'Add missing cases':

```cs
switch (((EntityHandle)h).Kind)
{
    case HandleKind.ModuleDefinition:
        break;
    case HandleKind.TypeReference:
        break;
    case HandleKind.TypeDefinition:
        break;
    // (...34 more)
}
```

But nothing is offered here:

```cs
return ((EntityHandle)h).Kind swi[||]tch
{
};
```

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes #37639

Looks like this:

image

Todo:

- [ ] tests

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jinujoseph @jnm2

**Comment by CyrusNajmabadi** [comment]
this is involved a bunch of refactoring to extract out common algorithms to share across switch-statements and switch-expressions.  That's the meat of this change.

File: src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        protected sealed override ITypeSymbol GetSwitchType(ISwitchOperation switchOperation)
+            => switchOperation.Value.Type;
+
+        protected sealed override ICollection<ISymbol> GetMissingEnumMembers(ISwitchOperation switchOperation)
+            => PopulateSwitchStatementHelpers.GetMissingEnumMembers(switchOperation);
+
+        protected sealed override TSwitchSyntax InsertSwitchArms(SyntaxGenerator generator, TSwitchSyntax switchNode, int insertLocation, List<TSwitchArmSyntax> newArms)
+            => (TSwitchSyntax)generator.InsertSwitchSections(switchNode, insertLocation, newArms);
+
+        protected sealed override TSwitchArmSyntax CreateDefaulSwitchArm(SyntaxGenerator generator, Compilation compilation)
```
Comment (jnm2) [inline src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs]: Typo `Defaul`

File: src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        protected sealed override int InsertPosition(ISwitchOperation switchStatement)
+        {
+            // If the last section has a default label, then we want to be above that.
+            // Otherwise, we just get inserted at the end.
+
+            var cases = switchStatement.Cases;
+            if (cases.Length > 0)
+            {
+                var lastCase = cases.Last();
```
Comment (jnm2) [inline src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs]: (option)

```cs
if (cases.LastOrDefault() is { } lastCase)
{
```

File: src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs]: fixed.

File: src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs]: dont' like :)

File: src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs [hunk 3]
```diff
@@ -115,6 +115,8 @@ internal static class IDEDiagnosticIds
 
         public const string UseSystemHashCode = "IDE0070";
 
+        public const string PopulateSwitchExpressionDiagnosticId = "IDE0072";
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs]: intentional gap as i have another PR that is adding IDE0071

**Comment by CyrusNajmabadi** [comment]
@jnm2 thanks for your help!
@JoeRobich This is ready for review.  Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks @JoeRobich and @jnm2 !

**Pull request merged**